### PR TITLE
Add dynamic output gpio example to demonstrate how to use AnyPin and Flex

### PR DESCRIPTION
**Pull request by Szybet**
### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
I added an example which shows how to use the AnyPin and Flex, because it took me too much time to figure out myself and I don't want for anyone else to waste time on this.

#### Testing
Measured the GPIOs states on the ESP32-C6-DevKitM-1 using an oscilloscope.

### CODE REVIEW

File: examples/src/bin/dynamic_output_gpio.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+//! - LED => GPIO0
+//! - LED => GPIO1
+//! - LED => GPIO2
+
+//% CHIPS: esp32c6
+
+#![no_std]
+#![no_main]
+
+use core::primitive::u32;
```
Comment (bugadani) [inline examples/src/bin/dynamic_output_gpio.rs:13]: Why is this necessary?

Comment (Szybet) [reply examples/src/bin/dynamic_output_gpio.rs]: I wanted to make a fancy random blinky out of that, but I decided to make it simpler instead.
Well idk how rust analyzer didn't catch this, but it's gone now

**Review by MabezDev — CHANGES_REQUESTED** [review]
I think this would be better as an improvement to the documentation: https://docs.esp-rs.org/esp-hal/esp-hal/0.19.0/esp32c6/esp_hal/gpio/struct.Flex.html, instead of an example. As an example it's basically the same as the blinky_erased example.

Would you mind changing this PR to improve the docs for `Flex` instead?

**Review by MabezDev — CHANGES_REQUESTED** [review]
I think this would be better as an improvement to the documentation: https://docs.esp-rs.org/esp-hal/esp-hal/0.19.0/esp32c6/esp_hal/gpio/struct.Flex.html, instead of an example. As an example it's basically the same as the blinky_erased example.

Would you mind changing this PR to improve the docs for `Flex` instead?

**Review by MabezDev — CHANGES_REQUESTED** [review]
I think this would be better as an improvement to the documentation: https://docs.esp-rs.org/esp-hal/esp-hal/0.19.0/esp32c6/esp_hal/gpio/struct.Flex.html, instead of an example. As an example it's basically the same as the blinky_erased example.

Would you mind changing this PR to improve the docs for `Flex` instead?

**Review by MabezDev — CHANGES_REQUESTED** [review]
I think this would be better as an improvement to the documentation: https://docs.esp-rs.org/esp-hal/esp-hal/0.19.0/esp32c6/esp_hal/gpio/struct.Flex.html, instead of an example. As an example it's basically the same as the blinky_erased example.

Would you mind changing this PR to improve the docs for `Flex` instead?

**Comment by Szybet** [comment]
> I think this would be better as an improvement to the documentation: https://docs.esp-rs.org/esp-hal/esp-hal/0.19.0/esp32c6/esp_hal/gpio/struct.Flex.html, instead of an example. As an example it's basically the same as the blinky_erased example.
> 
> Would you mind changing this PR to improve the docs for `Flex` instead?

The point of this was to make it easier to find this solution, adding it to the docs somewhere is not really a solution to what I'm trying to solve

**Comment by MabezDev** [comment]
> The point of this was to make it easier to find this solution, adding it to the docs somewhere is not really a solution to what I'm trying to solve

Maybe I'm missing something, but if the usage is in the docs (remember, we can add code snippets to the docs) it's probably more discoverable than it being in an example?

**Comment by Szybet** [comment]
> Maybe I'm missing something, but if the usage is in the docs (remember, we can add code snippets to the docs) it's probably more discoverable than it being in an example?

To find it in docs you need to know what you are looking for, while examples are more easily discoverable by just the description of what are they doing

**Comment by bjoernQ** [comment]
I agree with @MabezDev - maybe extending `blinky_erased_pins` would be an option but another example just for GPIO is somewhat "expensive"

Every example is built for every target in CI (which is run for every commit in every PR). Additionally, there is maintenance effort needed whenever something gets refactored

**Comment by Szybet** [comment]
Ok, I'm gonna look into it

**Comment by Szybet** [comment]
Do I just expand it by putting it into `Flex` and `AnyPin` and making a comment "You can use it as output now too"?

Also, what's with the name erased pins? I don't get it

**Comment by bjoernQ** [comment]
"erased pin" = type erased pin - things changed a lot since adding this example and the name probably isn't a great fit anymore

**Comment by bugadani** [comment]
Please note that the GPIO driver is under rework, so things **will** change around here.

**Comment by Szybet** [comment]
Sure, I will look into it once it's ready

Have you thought maybe about a subset of examples which wouldn't be compiled on every commit (to my knowledge they are compiled with regular cargo build, without release), just `cargo check`? This example could fit in it, the one from https://github.com/esp-rs/esp-hal/pull/2032 too.

Or just run the check just before merging a pull request? I understand why we should limit the amount of examples, but I think they are important

**Comment by bjoernQ** [comment]
Converted to draft for now - set it to Read-for-Review once it's done

**Comment by MabezDev** [comment]
Thanks for the PR!

We're going to be reworking our current example set soon in favour of more complete "mini-project" type of examples, therefore this PR doesn't really fit in that scheme.
